exercises/concept/story-name/.docs/introduction.md

```diff
@@ -1,27 +1,141 @@
+
+> This is why the `Html` package defines the `Html msg` type, with one type parameter named `msg`.
+> The elm architecture guides you towards creation of a custom type `Msg` describing all messages that your web page can trigger.
+> As a result, your code will be building views of the type `Html Msg` where `Msg` is the concrete type that you defined.
+> Newcommers are sometimes confused between errors mentioning `Html msg` (lowercase) and errors with `Html Msg`.
+> Now you know!
+> The difference is that the first one is generic over any `msg` type, while the second is specific to your `Msg` type.
 
 ## Type parameters in functions
 
-- as input
-- as output
-- as both input and output
-- parametric polymorphism VS ad hoc polymorphism VS subtyping polymorphism
+Since types can be generic, so can functions.
+Functions handling generic types can be writen, such as the `List.length` function that we mentioned at the beginning.
+
+```elm
+length : List a -> Int
+length list =
+    case list of
+        [] ->
+            0
+        _ :: restOfTheList ->
+            1 + length restOfTheList
+```
+
+Type parameters can appear both in the inputs and in the output of a function, and genericity enabled by type parameters is usually called **parametric polymorphism**.
+This is in contrast with ad hoc polymorphism also known as function overloading, where the same function name can be used on multiple implementation where the types differ.
+This also contrasts with subtype polymorphism, where types can be defined in a hierarchy, and functions defined on a type can also be called with arguments of its subtypes.
+In elm, the only available mechanism for genericity is parametric polymorphism.
+
+## Enforcing compile-time constraints with phantom types
+
+In most cases, type parameters are introduced to enable genericity.
+This is what happens when the type parameter is present on both sides of the equal sign `=` in the type definition, as in our `Maybe` definition.
 
-## Higher order functions
+```elm
+-- the type parameter 'a' is present both on the left side
+-- and on the right side of the equal sign '='.
+type Maybe a = Nothing | Just a
+```
+
+We say that the type parameter `a` is bound to some data inside the type definition.
+In certain cases however, the type parameter only appears on the left side of the equation.
+
+```elm
+type Distance unit = Distance Float
+```
+
+In the definition of `Distance` above, `unit` is a free type parameter, not bound to any data in the type.
+We can also call this technique **phantom types**.
+It is surprisingly useful when we want to enforce constraints at compile type.
+For example, we want to make sure that we only add distances of the same unit.
+
+```elm
+-- The module does not expose the internals of Distance,
+-- but instead two constructor functions: meter and foot.
+-- We do not want to expose 'Distance(..)' because we do not want
+-- people outside this module to access the internals of Distance.
```

Comment:
I might word this bit slightly differently. The main thing with opaque types is that you can't create an instance of a type directly, you have to use one of the exported creator functions.